oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthSessionAuthorizer.java

```diff
@@ -0,0 +1,159 @@
+
+    private kafka.security.auth.Authorizer delegate;
+
+    @Override
+    public void configure(java.util.Map<String, ?> configs) {
+        String className = (String) configs.get(ServerConfig.STRIMZI_AUTHORIZER_DELEGATE_CLASS_NAME);
+
+        if (className != null) {
+            try {
+                Class delegateClass = Thread.currentThread().getContextClassLoader().loadClass(className);
+                if (!kafka.security.auth.Authorizer.class.isAssignableFrom(delegateClass)) {
+                    throw new IllegalArgumentException("The class specified by " + ServerConfig.STRIMZI_AUTHORIZER_DELEGATE_CLASS_NAME + " is not an instance of kafka.security.auth.Authorizer");
+                }
+
+                delegate = (kafka.security.auth.Authorizer) delegateClass.<kafka.security.auth.Authorizer>newInstance();
+
+                // Configure the delegate
+                delegate.configure(configs);
+
+            } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+                throw new RuntimeException("Failed to instantiate and configure the delegate authorizer: " + className, e);
+            }
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Configured OAuthSessionAuthorizer:"
+                    + "\n    " + ServerConfig.STRIMZI_AUTHORIZER_DELEGATE_CLASS_NAME + ": " + className);
+        }
+    }
+
+    @Override
+    public boolean authorize(RequestChannel.Session session, kafka.security.auth.Operation operation, kafka.security.auth.Resource resource) {
+
+        KafkaPrincipal principal = session.principal();
+
+        if (!(principal instanceof OAuthKafkaPrincipal)) {
+            // If user wasn't authenticated over OAuth, there's nothing for us to check
+            if (delegate != null) {
+                return delegate.authorize(session, operation, resource);
+            } else {
+                if (GRANT_LOG.isDebugEnabled()) {
+                    GRANT_LOG.debug("Authorization GRANTED - no access token: " + session.principal() + ", operation: " + operation + ", resource: " + resource);
+                }
+                return true;
+            }
+        }
+
+        BearerTokenWithPayload token = ((OAuthKafkaPrincipal) principal).getJwt();
+
+        if (denyIfTokenInvalid(token)) {
+            return false;
+        }
+
+        if (delegate == null) {
+            if (GRANT_LOG.isDebugEnabled()) {
+                GRANT_LOG.debug("Authorization GRANTED - access token still valid: " + session.principal() + ", operation: " + operation + ", resource: " + resource + ", token: " + mask(token.value()));
+            }
+            return true;
+        }
+
+        return delegate.authorize(session, operation, resource);
+    }
+
+    private boolean denyIfTokenInvalid(BearerTokenWithPayload token) {
+        if (token.lifetimeMs() <= System.currentTimeMillis()) {
```

Comment:
Why is it called `lifetimeMs` (which sounds like a duration) rather than `expiryTime` (or `expiryTimeOffset` or something if you want to make the datum clearer).

Reply:
That's the question for Kafka developers: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerToken.java#L76-L86

Reply:
That's a fair point!